src/backend/metal/src/command.rs

```diff
@@ -1,477 +1,634 @@
-
-        // TODO: blend/stencil
+    fn bind_compute_pipeline(&mut self, pipeline: &native::ComputePipeline) {
+        unimplemented!()
     }
-
-    #[allow(unused_variables)]
-    fn copy_buffer(&mut self, src: Buffer, dst: Buffer,
-                   src_offset_bytes: usize, dst_offset_bytes: usize,
-                   size_bytes: usize) {
+    
+    fn dispatch(&mut self, x: u32, y: u32, z: u32) {
         unimplemented!()
     }
-
-    #[allow(unused_variables)]
-    fn copy_buffer_to_texture(&mut self, src: Buffer, src_offset_bytes: usize,
-                              dst: Texture,
-                              kind: texture::Kind,
-                              face: Option<texture::CubeFace>,
-                              img: texture::RawImageInfo) {
+    
+    fn dispatch_indirect(&mut self, buffer: &native::Buffer, offset: u64) {
         unimplemented!()
     }
-
-    #[allow(unused_variables)]
-    fn copy_texture_to_buffer(&mut self,
-                              src: Texture,
-                              kind: texture::Kind,
-                              face: Option<texture::CubeFace>,
-                              img: texture::RawImageInfo,
-                              dst: Buffer, dst_offset_bytes: usize) {
+    
+    fn copy_buffer(
+        &mut self,
+        src: &native::Buffer,
+        dst: &native::Buffer,
+        regions: &[BufferCopy],
+    ) {
         unimplemented!()
     }
-
-    fn update_buffer(&mut self, buf: Buffer, data: &[u8], offset: usize) {
-        use map::{map_buffer_usage};
-
-        debug_assert!(!unsafe { *(*(buf.0).0) }.is_null(), "Buffer must be non-nil");
-
-        let b = unsafe { *(buf.0).0 };
-
-        // we can create a new buffer to avoid synchronization when:
-        //   * buffer is write-only
-        //   * we are replacing all the contents of the buffer
-        //
-        // TODO(fkaa): maybe have an upper limit to prevent from thrashing?
-        if offset == 0 && data.len() == b.length() as usize {
+    
+    fn copy_image(
+        &mut self,
+        src: &native::Image,
+        src_layout: ImageLayout,
+        dst: &native::Image,
+        dst_layout: ImageLayout,
+        regions: &[ImageCopy],
+    ) {
+        unimplemented!()
+    }
+    
+    fn copy_buffer_to_image(
+        &mut self,
+        src: &native::Buffer,
+        dst: &native::Image,
+        dst_layout: ImageLayout,
+        regions: &[BufferImageCopy],
+    ) {
+        let encoder = self.encode_blit();
+        let extent = unsafe { MTLSize {
+            width: dst.0.width(),
+            height: dst.0.height(),
+            depth: dst.0.depth(),
+        }};
+        // FIXME: layout
+
+        for region in regions {
+            let image_offset = &region.image_offset;
+
+            // TODO multiple layers
+            assert!((region.image_subresource.1).len() == 1, "multiple layer copies not implemented");
```

Comment:
We can't loop over all layers?

Reply:
I think for array/cube textures extent.depth might actually map to layer count (otherwise sourceBytesPerImage doesn't make a ton of sense). I think it might be better to wait until we have a test case before we try to implement it that way though.